src/rqt_msg/messages_widget.py

```diff
@@ -211,34 +214,46 @@ def _rightclick_menu(self, event):
             self._logger.debug('_rightclick_menu selected={}'.format(selected))
             selected_type = selected[1].data()
 
-            if selected_type.find('[') >= 0:
-                selected_type = selected_type[:selected_type.find('[')]
+            # We strip any array information for loading the python classes
+            selected_type_bare = selected_type
+            if selected_type_bare.find('[') >= 0:
+                selected_type_bare = selected_type_bare[:selected_type_bare.find('[')]
+            # We use the number of '/' to determine of the selected type is a msg, action, srv,
+            # or primitive type.
+            # NOTE (mlautman - 2/4/19) this heuristic seems brittle and should be removed
+            selected_type_bare_tokens_len = len(selected_type_bare.split('/'))
+
+            # We only want the base class so we transform eg. pkg1/my_srv/Request -> pkg1/my_srv
+            if selected_type_bare_tokens_len > 2:
+                selected_type_bare = "/".join(selected_type_bare.split('/')[:2])
 
-            # TODO(mlautman):
-            #   implement get_msg_text like functionality like what is available in ROS1
             browsetext = None
 
-            if (self._mode == message_helpers.MSG_MODE):
-                if is_primitive_type(selected_type):
-                    browsetext = selected_type
-                else:
-                    msg_class = get_message_class(selected_type)
-                    browsetext = get_message_text_from_class(msg_class)
+            # If the type does not have '/'s then we treat it as a primitive type
+            if selected_type_bare_tokens_len == 1:
+                browsetext = selected_type
 
-            elif self._mode == message_helpers.SRV_MODE:
-                if is_primitive_type(selected_type):
-                    browsetext = selected_type
-                else:
-                    msg_class = get_service_class(selected_type)
-                    browsetext = get_service_text_from_class(msg_class)
+            # if the type has a single '/' then we treat it as a msg type
+            elif selected_type_bare_tokens_len == 2:
+                msg_class = get_message_class(selected_type_bare)
+                browsetext = get_message_text_from_class(msg_class)
 
-            elif self._mode == message_helpers.ACTION_MODE:
-                if is_primitive_type(selected_type):
-                    browsetext = selected_type
-                else:
-                    msg_class = get_action_class(selected_type)
+            # If the type has two '/'s then we treat it as a srv or action type
+            elif selected_type_bare_tokens_len == 3:
```

Comment:
The same comment as on the [rqt PR](https://github.com/ros-visualization/rqt/pull/188#discussion_r253130114) applies here as well. This should not be merged as is.

Reply:
To address your point from the rqt PR, we would never see `<pkgname>/<srvname>` since we explicitly append `/Request` and `/Response` before constructing the message tree for each of them in the service browser. Since this is a private function that should only be used from within the class, we can use this contextual information to determine if the user clicked on the service `Request`/`Response` or if they clicked on one of the children msgs or primitive types in the `MessageTreeModel`.

Reply:
Oh, I see, this is a slightly different use than in the rqt PR then. I don't notice that. Thanks for clarifying. Good as is then :+1:

Reply:
@dirk-thomas I noticed that you never gave @brawner or myself permission to release rqt packages. Can you take care of releasing the new changes to rqt? I took care of the steps up to `bloom-release`

Reply:
Done: ros/rosdistro#20175

In the future please ask on the corresponding repo instead.